docs/unification.md

```diff
@@ -0,0 +1,344 @@
+each subterm according to one of three categories:
+
+1. Elements
+2. Variables
+3. Functions (other than `.Map`, `_Map_`, and `|->`)
+
+This process may yield the value \bottom for either map, in which case, fail.
+
+First, we remove all subterms present in both maps from each map.
+
+We then consider 2 cases:
+
+1. Either map still contains functions. In this case, constrain with an
+   equality between the two maps.
+2. Both maps contain no functions. In this case, apply AC unification to the
+   two maps.
+
+### AC Unification 
+
+First, we consider the following special cases:
+
+* If both maps are empty, discharge.
+* If one map has 0 elements and 1 variable, treat it as a variable pattern
+  instead, and perform variable abstraction on the other pattern, then follow
+the procedure for AC binding.
+* If both maps have 1 element and 0 variables, decompose with the keys and
+  values.
+* Otherwise, if one map is empty and the other has 0 elements, bind each
+  variable of the other map to the empty map.
+* Otherwise, if one map is empty (and the other has elements), fail.
+* Otherwise, perform variable abstraction on the elements of both maps and
+  decompose with the AC equation and free equations that result.
+
+### Variable Abstraction
+
+Variable abstraction replaces each element in one or both map patterns with
+a new fresh variable and generates a free equality equating that variable with
+the element it corresponds to. It then is left with an equality that consists
+purely of map concatenation and variables, and adds it to the AC equations to
+solve.
+
+### AC Binding
+
+If one term is a variable X and the other is `_Map_(s1, ..., sn)` where si is a
+variable, we consider the following cases:
+
+* If X is free, bind X to the other term.
+* If X is bound to a variable, bind the variable it is bound to to the other
+  term.
+* Otherwise, add the AC equation between the binding of X and the other term to
+  the AC equations to solve.
+
+### AC Subproblems
+
+After the priority queue is empty, we are left with one AC subproblem per map
+or set sort. We then solve each one according to the following algorithm, which
+returns a list of disjunct solutions:
+
+First, we replace each variable with its binding in the substitution until this
+can no longer occur.
+
+We are left with a system of equations P == s1 = s1' /\ ... /\ sp = sp' where
+aeach si and si' is of the form `_Map_(t1, ..., tk)` for some k > 1 where each
+ti is a variable. We define u1, ..., un to be the unique variables in the
+system of equations, and associate each ui an integer variable xi. We say a
+variable xi is *constrained* if it came from a variable abstraction of a map
+element, and *unconstrained* otherwise. We then convert P into the following
+system of linear Diophantine equations:
+
+d(u1, t1=t1')x1 + ... + d(un, t1=t1')xn = 0
+.
+.
+.
+d(u1, tp = tp')x1 + ... + d(un, tp=tp')xn = 0
+
+Where d(ui, ti = ti') is the number of occurrences of ui in ti minus the number
```

Comment:
Maybe? But really I am just pretty faithfully rendering the algorithm provided in the Boudet paper. I'd rather not tinker with this since it already works pretty well, even though you are correct that the algorithm doesn't deal explicitly with nilpotence and thus could potentially be simplified somewhat.